### PR TITLE
feat: Introduce operational metrics in OpenTelemetry Collector

### PR DESCRIPTION
**Pull request by SpiritZhou**
Refactor the current Prometheus.go file by adding a middleware metriccollector.go to handle initialization and metrics collection for both Prometheus and OpenTelemetry.

About e2e test:
1. Add opentelemetry install in setup_test.go. This opentelemetry has promethues exporter.
2. Add env and flag when KEDA is running in e2e test env.
3. Fecth metric from opentelemetry  through promethues port to test 

### Checklist

- [X] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [X] Tests have been added
- [X] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [X] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Relates to https://github.com/kedacore/keda/issues/3078
Relates to https://github.com/kedacore/keda-docs/pull/1215

### CODE REVIEW

File: tests/opentelemetry_setup/otlp.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+mode: deployment
```
Comment (JorTurFer) [inline tests/opentelemetry_setup/otlp.yml:1]: Even though having a file is good enough, could you place these required files as constants inside the test case? It's more consistent with current approach and also helps identifying where are used

File: Makefile [hunk 2]
```diff
@@ -285,6 +285,10 @@ deploy: install ## Deploy controller to the K8s cluster specified in ~/.kube/con
 		cd config/service_account && \
 		$(KUSTOMIZE) edit add annotation --force cloud.google.com/workload-identity-provider:${GCP_WI_PROVIDER} cloud.google.com/service-account-email:${TF_GCP_SA_EMAIL} cloud.google.com/gcloud-run-as-user:${NON_ROOT_USER_ID}; \
 	fi
+	if [ "$(OEPNTELEMETRY_RUN_TESTS)" = true ]; then \
+		cd config/e2e && \
+		$(KUSTOMIZE) edit add component ../../tests/opentelemetry_setup/keda_setup/; \
```
Comment (JorTurFer) [inline Makefile:290]: OTEL tests should be executed always, so the collector has to be installed always :)

**Review by JorTurFer — COMMENTED** [review]
This is an awesome improvement!! 💪 
I have left some comments inline, but apart from them:

As testing otel is a must, I'd move the Kustomization to config/e2e because it's the Kustomization file that we use for deploying e2e tests (for releases we use config/default). It makes easier to maintain the files and also ensures that always is deployed

We also have to document this new option in keda-docs 🙏

File: CHANGELOG.md [hunk 3]
```diff
@@ -56,6 +56,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 - **General**: Add ScaledObject/ScaledJob names to output of `kubectl get triggerauthentication/clustertriggerauthentication` ([#796](https://github.com/kedacore/keda/issues/796))
 - **General**: Add standalone CRD generation to release workflow ([#2726](https://github.com/kedacore/keda/issues/2726))
 - **General**: Adding a changelog validating script to check for formatting and order ([#3190](https://github.com/kedacore/keda/issues/3190))
+- **General**: Introduce opentelemetry metric collect ([#3078](https://github.com/kedacore/keda/issues/3078))
```
Comment (tomkerkhove) [inline CHANGELOG.md:59]:
```suggestion
- **General**: Introduce operational metrics in OpenTelemetry Collector ([#3078](https://github.com/kedacore/keda/issues/3078))
```

**Comment by tomkerkhove** [comment]
We also need to expose this in our Helm chart but that's a separate PR as well

File: pkg/metricscollector/opentelemetry.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+func initCounter() {
+	var err error
+	msg := "create opentelemetry counter failed"
+
+	otErrorsTotalCounter, err = meter.Int64Counter("scaler.errors.total", api.WithDescription("Total number of errors for all scalers"))
+	if err != nil {
+		otLog.Error(err, msg)
+	}
+
+	otScalerErrorsCounter, err = meter.Int64Counter("scaler.errors", api.WithDescription("Number of scaler errors"))
```
Comment (pragmaticivan) [inline pkg/metricscollector/opentelemetry.go:68]: Is there any difference other than label/attribute presence in `scaler.errors.total` vs `scaler.errors` ? afaik the `sum()` function anywhere would still produce the same result, as long you don't filter it by any attribute.

Comment (pragmaticivan) [reply pkg/metricscollector/opentelemetry.go:68]: ref: https://github.com/kedacore/keda/pull/4860/files#diff-1b98f9b287bb8783cd25e2877e1897ecfca1e2a73c6852339679014d211c51ecR170-R172

The same applies to the Prometheus implementation. Got confused with that one for a sec.

Comment (SpiritZhou) [reply pkg/metricscollector/opentelemetry.go:68]: I think that is for easy analytic? Do you know the main reason of these metrics? @tomkerkhove @JorTurFer

Comment (JorTurFer) [reply pkg/metricscollector/opentelemetry.go:68]: Yeah, we added them in the past to make the things easier for end users but it's true that probably it's worse for end users and they have to store another time series when Prometheus can just sum the values. IMO we can deprecate those repeated metrics if we documented it well, but we can't remove them directly and I also think that prometheus and otel should expose the same metrics.
Any thought @zroubalik @tomkerkhove ?

Comment (zroubalik) [reply pkg/metricscollector/opentelemetry.go:68]: I would deprecate them in Prometheus and if possible not even introduce in otel.

Comment (JorTurFer) [reply pkg/metricscollector/opentelemetry.go:68]: @tomkerkhove ?

Comment (JorTurFer) [reply pkg/metricscollector/opentelemetry.go:68]: Does sending the data in multiple ways make sense? Ingesting and processing telemetry means a cost, and we are generating multiple time series for things that end users can aggregate just in time if they need it.

**Review by JorTurFer — COMMENTED** [review]
I have reopened the comment about generating aggregated metrics or not. Apart from that, this PR looks awesome and LGTM!
There are also a comment on docs about metric names that maybe requires any change here

File: apis/keda/v1alpha1/scaledobject_webhook.go [hunk 5]
```diff
@@ -33,7 +33,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 
-	prommetrics "github.com/kedacore/keda/v2/pkg/prommetrics/webhook"
+	prommetrics "github.com/kedacore/keda/v2/pkg/metricscollector/webhook"
```
Comment (zroubalik) [inline apis/keda/v1alpha1/scaledobject_webhook.go:36]:
```suggestion
	metricscollector "github.com/kedacore/keda/v2/pkg/metricscollector/webhook"
```

or something like this^

File: pkg/metricscollector/opentelemetry.go (hunk 4 above)
Comment (zroubalik) [reply pkg/metricscollector/opentelemetry.go:68]: @tomkerkhove FYI^

File: CHANGELOG.md [hunk 6]
```diff
@@ -59,6 +59,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 - **General**: Add ScaledObject/ScaledJob names to output of `kubectl get triggerauthentication/clustertriggerauthentication` ([#796](https://github.com/kedacore/keda/issues/796))
 - **General**: Add standalone CRD generation to release workflow ([#2726](https://github.com/kedacore/keda/issues/2726))
 - **General**: Adding a changelog validating script to check for formatting and order ([#3190](https://github.com/kedacore/keda/issues/3190))
+- **General**: Introduce operational metrics in OpenTelemetry Collector ([#3078](https://github.com/kedacore/keda/issues/3078))
```
Comment (zroubalik) [inline CHANGELOG.md:62]: Move this to the `### New` section please. It is a great feature and it deserves to be promoted more :)

**Review by zroubalik — COMMENTED** [review]
Looking good, just a nit in Changelog and one import. 

Also we need to answer the outstanding issue on `total errors`

File: pkg/metricscollector/opentelemetry.go (hunk 4 above)
Comment (tomkerkhove) [reply pkg/metricscollector/opentelemetry.go:68]: If we still push the raw `scaler.errors` metric and not remove it then I'm fine

Comment (JorTurFer) [reply pkg/metricscollector/opentelemetry.go:68]: Yeah, the proposal is not losing any info but dropping aggregated time series which can be calculated on demand in the backend

File: tests/sequential/opentelemetry_metrics/opentelemetry_metrics_test.go [hunk 7]
```diff
@@ -0,0 +1,712 @@
...
+var (
+	testNamespace             = fmt.Sprintf("%s-ns", testName)
+	deploymentName            = fmt.Sprintf("%s-deployment", testName)
+	monitoredDeploymentName   = fmt.Sprintf("%s-monitored", testName)
+	scaledObjectName          = fmt.Sprintf("%s-so", testName)
+	wrongScaledObjectName     = fmt.Sprintf("%s-wrong", testName)
+	wrongScalerName           = fmt.Sprintf("%s-wrong-scaler", testName)
+	cronScaledJobName         = fmt.Sprintf("%s-cron-sj", testName)
+	clientName                = fmt.Sprintf("%s-client", testName)
+	kedaOperatorPrometheusURL = "http://opentelemetry-collector.default.svc.cluster.local:8889/metrics"
```
Comment (JorTurFer) [inline tests/sequential/opentelemetry_metrics/opentelemetry_metrics_test.go:41]: Could you rename this var? I thought that you were using the same endpoint as for promtheus tests.
Something like `otelBackendPrometheusURL`

**Review by zroubalik — COMMENTED** [review]
@SpiritZhou do you think you can address the last few points? We plan to cut the release tomorrow (Wednesday).
Also, we need to update Helm Charts as well.

**Comment by SpiritZhou** [comment]
> @SpiritZhou do you think you can address the last few points? We plan to cut the release tomorrow (Wednesday). Also, we need to update Helm Charts as well.

@zroubalik There are still two parts I don't see the decision. One is should we deprecate "scaler.errors.total" in this PR (https://github.com/kedacore/keda/pull/4860#discussion_r1328369583). Another one is should we add prefix "keda." in otel metric name (https://github.com/kedacore/keda-docs/pull/1215#discussion_r1334803007). Other points I can update soon.

File: pkg/metricscollector/opentelemetry.go (hunk 4 above)
Comment (tomkerkhove) [reply pkg/metricscollector/opentelemetry.go:68]: > If we still push the raw `scaler.errors` metric and not remove it then I'm fine

@SpiritZhou if we still have  the raw metric than you can drop the aggregated one

**Comment by tomkerkhove** [comment]
@JorTurFer @zroubalik in my opinion prefixing with `keda` is a good idea but don't have a strong opinion

**Comment by zroubalik** [comment]
> @JorTurFer @zroubalik in my opinion prefixing with `keda` is a good idea since we own the metric names pushed to OTEL Collector but don't have a strong opinion

Agree with @tomkerkhove on this, drop total errors and prefix fix `keda` please.

**Comment by JorTurFer** [comment]
+1

**Comment by SpiritZhou** [comment]
@zroubalik @JorTurFer Just updated the code. Hope it can meet the ddl.

**Review by zroubalik — COMMENTED** [review]
Unfortunately, prometheus tests is failing:
```
prometheus_metrics_test.go:336: --- testing scaled object errors ---
    helper.go:567: Deleting template: scaledObjectTemplate
    helper.go:5***9: Applying template: wrongScaledObjectTemplate
    prometheus_metrics_test.go:35***: 
        	Error Trace:	/__w/keda/keda/tests/sequential/prometheus_metrics/prometheus_metrics_test.go:35***
        	            				/__w/keda/keda/tests/sequential/prometheus_metrics/prometheus_metrics_test.go:***67
        	Error:      	Should not be: 0
        	Test:       	TestPrometheusMetrics
```

**Review by zroubalik — COMMENTED** [review]
@SpiritZhou one more thing, could you please make it an experimental feature? Marking it so in the docs and changelog? 

There's nothing wrong with the feature itself. Just for us to be able to change the metrics or nomenclature if needed without the need to keep backwards compatibility.

**Comment by SpiritZhou** [comment]
@zroubalik I fixed the test and update the "experimental feature". Could you help to review them again?

**Review by zroubalik — APPROVED** [review]
LGTM, thanks for the quick fix!

**Comment by tomkerkhove** [comment]
Talked to @SpiritZhou and he'll start Helm chart PR, if he can't finish it today I'll take over tomorrow. That should allow us to merge this one in time

**Comment by SpiritZhou** [comment]
It is strange. The prometheus_metrics_test can be passed in my local minikube env...

**Comment by zroubalik** [comment]
> @zroubalik @JorTurFer @tomkerkhove The failed e2e test may not be caused by OpenTelemetry PR. I found it happened in this PR before (https://github.com/kedacore/keda/actions/runs/6308971553). And I also found that the correct metric value can be retrieved after around 15s. I don't have any idea why it becomes slow.

Right, it is failing on `main` as well. So it is not relevant to this PR. You can ignore it in this PR, great work, thanks!

**Comment by tomkerkhove** [comment]
I will pick up Helm chart later today

**Pull request merged**